nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py

```diff
@@ -0,0 +1,252 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+from omegaconf.dictconfig import DictConfig
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.common.parts import adapter_modules
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_prompt_learning_model import (
+    MegatronGPTPromptLearningModel,
+)
+from nemo.collections.nlp.modules.common import VirtualPromptStyle
+from nemo.collections.nlp.modules.common.megatron.utils import average_losses_across_data_parallel_group
+from nemo.collections.nlp.parts.utils_funcs import get_last_rank
+from nemo.core.classes.mixins import adapter_mixins
+from nemo.utils import logging
+
+
+class MegatronGPTAdapterLearningModel(MegatronGPTPromptLearningModel):
+    """
+    MegatronGPTAdapterLearningModel is a model that combines a base model (GPTModel) with a adapters.
+    This class only supports the canonical Adapter training described in Houlsby et al. (https://arxiv.org/pdf/1902.00751.pdf)
+
+    Two adapter's are inserted into each Transformer layer in the base GPT Model.
+
+    It is assumed that these set of adapters will then be trained for a specific task.
+    Once trained, the adapter weights will be saved and can be re-loaded 
+    and infused into the same GPT Model for inference. 
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+
+        self.adapter_name_keys = ['adapter_1', 'adapter_2']
+        frozen_model_cfg = MegatronGPTModel.restore_from(
+            cfg.get('language_model_path'), trainer=trainer, return_config=True
+        )
+        for _, layer in self.frozen_model.named_modules():
+            if hasattr(layer, 'activations_checkpoint_method'):
+                layer.activations_checkpoint_method = (
+                    None  # (@adithyare) adapter learning does not support activations checkpointing atm.
+                )
+            if hasattr(layer, 'scale_mask_softmax'):
+                layer.scale_mask_softmax.scaled_masked_softmax_fusion = False
+
+        logging.info(f'Before adding adapters:\n{self.frozen_model.summarize()}')
+        self.frozen_model.freeze()
+        for _, module in self.frozen_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    module.add_adapter(
+                        name=adapter_key,
+                        cfg=adapter_modules.LinearAdapterConfig(
+                            in_features=frozen_model_cfg.hidden_size,
+                            dim=cfg.adapter_tuning.adapter_dim,
+                            norm_position='post',
```

Comment:
Norm pos should default to "pre" to be accurate to papers.

Reply:
I'll make this configurable per your suggestion. My experiments suggest post is better in some cases.